base/compiler/typeinfer.jl

```diff
@@ -569,49 +570,51 @@ function typeinf_ext(mi::MethodInstance, params::Params)
         end
     end
     mi.inInference = true
-    frame = InferenceState(InferenceResult(mi), #=cached=#true, params)
+    frame = InferenceState(InferenceResult(mi), #=cached=#true, interp)
     frame === nothing && return nothing
-    typeinf(frame)
+    typeinf(interp, frame)
     ccall(:jl_typeinf_end, Cvoid, ())
     frame.src.inferred || return nothing
     return frame.src
 end
 
 # compute (and cache) an inferred AST and return the inferred return type
-function typeinf_type(method::Method, @nospecialize(atypes), sparams::SimpleVector, params::Params)
+function typeinf_type(interp::AbstractInterpreter, method::Method, @nospecialize(atypes), sparams::SimpleVector)
     if contains_is(unwrap_unionall(atypes).parameters, Union{})
         return Union{} # don't ask: it does weird and unnecessary things, if it occurs during bootstrap
     end
     mi = specialize_method(method, atypes, sparams)::MethodInstance
     for i = 1:2 # test-and-lock-and-test
         i == 2 && ccall(:jl_typeinf_begin, Cvoid, ())
-        code = inf_for_methodinstance(mi, params.world)
+        code = inf_for_methodinstance(interp, mi, get_world_counter(interp))
         if code isa CodeInstance
             # see if this rettype already exists in the cache
             i == 2 && ccall(:jl_typeinf_end, Cvoid, ())
             return code.rettype
         end
     end
     frame = InferenceResult(mi)
-    typeinf(frame, true, params)
+    typeinf(interp, frame, true)
     ccall(:jl_typeinf_end, Cvoid, ())
     frame.result isa InferenceState && return nothing
     return widenconst(frame.result)
 end
 
-@timeit function typeinf_ext(linfo::MethodInstance, world::UInt)
+# This is a bridge for the C code calling `jl_typinf_func()`
+typeinf_ext(mi::MethodInstance, world::UInt) = typeinf_ext(NativeInterpreter(world), mi, world)
+function typeinf_ext(interp::AbstractInterpreter, linfo::MethodInstance, world::UInt)
```

Comment:
same comment here, I don't think I want this to necessarily be an argument.

Reply:
How should we disambiguate this `typeinf_ext(::AbstractInterpreter, ::MethodInstance)` from the one above?

Reply:
I'm a bit confused what the difference between these is supposed to be @JeffBezanson?